### PR TITLE
Refactor parquet lz4 and lzo decompression in to dwio common package

### PR DESCRIPTION
**Pull request by nmahadevuni**

### CODE REVIEW

**Comment by nmahadevuni** [comment]
@yingsu00 Please review this.

File: velox/dwio/common/compression/Compression.cpp [hunk 1]
```diff
@@ -194,14 +195,139 @@ uint64_t ZlibDecompressor::decompress(
...
+      inputPtr += compressedLength;
+      inputLength -= compressedLength;
+      decompressedBlockLength -= decompressedSize;
+      decompressedTotalLength += decompressedSize;
+    } while (decompressedBlockLength > 0);
+  }
+
+  VELOX_CHECK_EQ(decompressedTotalLength, uncompressedSize);
+
+  return decompressedTotalLength - uncompressedSize;
```
[the remark this answers is not in the document]
Comment (yingsu00) [reply velox/dwio/common/compression/Compression.cpp:315]: I don't see any existing tests to test PagedInputStream and compression/decompressions. I think we should add some, maybe in a separate PR. cc @Yuhta

File: velox/dwio/parquet/reader/PageReader.cpp [hunk 2]
```diff
@@ -247,7 +138,7 @@ const char* FOLLY_NONNULL PageReader::decompressData(
           std::move(inputStream),
           uncompressedSize,
           pool_,
-          getParquetDecompressionOptions(),
+          getParquetDecompressionOptions(ThriftCodecToCompressionKind(codec_)),
```
[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/parquet/reader/PageReader.cpp:141]: Made it a static method.

File: velox/dwio/common/compression/Compression.cpp [hunk 3]
```diff
@@ -194,14 +195,139 @@ uint64_t ZlibDecompressor::decompress(
...
+      if (inputLength < sizeof(uint32_t)) {
+        VELOX_FAIL(
+            "{} decompression failed, input len is to small: {}",
+            kind_,
+            inputLength)
+      }
+      // Read the length of the next lz4/lzo compressed block.
+      uint32_t compressedLength =
+          folly::Endian::big(folly::loadUnaligned<uint32_t>(inputPtr));
+      inputPtr += dwio::common::INT_BYTE_SIZE;
```
[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/common/compression/Compression.cpp:279]: This code was already present. Changed it to use INT_BYTE_SIZE in both places.

File: velox/dwio/common/compression/Compression.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/common/compression/Compression.cpp:315]: For streaming decompressors it should be the current decompressedTotalLength. For other all-at-once decompressors, it should be the difference with the input uncompressedSize and it should be zero, which will be checked by the caller.

File: velox/dwio/common/compression/Compression.cpp [hunk 4]
```diff
@@ -194,14 +195,139 @@ uint64_t ZlibDecompressor::decompress(
...
+      // Decompress this block.
+      remainingOutputSize = uncompressedSize - decompressedTotalLength;
+      uint64_t decompressedSize = -1;
+      decompressedSize = decompressInternal(
+          inputPtr,
+          static_cast<int32_t>(compressedLength),
+          outPtr,
+          static_cast<int32_t>(remainingOutputSize));
+
+      VELOX_CHECK_EQ(decompressedSize, remainingOutputSize);
```
[the remark this answers is not in the document]
Comment (nmahadevuni) [reply velox/dwio/common/compression/Compression.cpp:303]: Changed to DWIO_ENSURE_EQ to keep in consistent in the file.

**Comment by nmahadevuni** [comment]
@Yuhta @xiaoxmeng Can you please take a look at this?

**Comment by nmahadevuni** [comment]
@Yuhta @xiaoxmeng Please review this PR.

**Comment by nmahadevuni** [comment]
@yingsu00 Addressed all your comments, please review.

File: velox/dwio/common/compression/Compression.cpp [hunk 5]
```diff
@@ -194,14 +195,158 @@ uint64_t ZlibDecompressor::decompress(
...
+  DWIO_ENSURE_EQ(
+      decompressedTotalSize,
+      uncompressedSize,
+      "{} decompression failed, decompressedTotalSize is not equal to uncompressedSize, "
+      "decompressedTotalSize: {}, uncompressedSize: {}",
+      kind_,
+      decompressedTotalSize,
+      uncompressedSize);
+
+  return decompressedTotalSize - uncompressedSize;
```
Comment (yingsu00) [inline velox/dwio/common/compression/Compression.cpp:334]: > For streaming decompressors it should be the current decompressedTotalLength. For other all-at-once decompressors, it should be the difference with the input uncompressedSize and it should be zero, which will be checked by the caller.

By "all-at-once decompressors", did you mean the framed LzoAndLz4Decompressor? Why should it return 0 always? The DWIO_ENSURE_EQ check ensures that decompressedTotalSize - uncompressedSize = 0 and therefore this decompress() always returns 0 for framed LZ4 and LZO decompression. As far as I understand, `Decompressor::decompress()` function should return the number of bytes written to the dest buffer no matter what compression type it is. The PagedInputStream has this callsite:
```
outputBufferLength_ = decompressor_->decompress(
        input,
        remainingLength_,
        outputBuffer_->data(),
        outputBuffer_->capacity());
    *data = outputBuffer_->data();
    *size = static_cast<int32_t>(outputBufferLength_);
    outputBufferPtr_ = outputBuffer_->data() + outputBufferLength_;
```
This is the only callsite, and it suggest it returns the output buffer length, which should not be 0 for most cases. Why would `outputBufferPtr_` be advanced with 0 bytes always?

Comment (nmahadevuni) [reply velox/dwio/common/compression/Compression.cpp:334]: Yes, you are right, I will have a relook at it. I had seen this callsite and verified that tests were all running fine. But right now, it seems parquet compression code is broken (https://github.com/facebookincubator/velox/issues/6890) and I will have a relook at this after a fix for this is introduced.

**Comment by nmahadevuni** [comment]
@yingsu00 The PR for LZ4 codec move from Arrow to Velox was merged today. I rebased this PR again. Please review. The failed linux-adapters requires a rerun. This is not related to the code changes.

File: velox/dwio/parquet/reader/PageReader.h [hunk 6]
```diff
@@ -68,8 +68,8 @@ class PageReader {
         chunkSize_(chunkSize),
         nullConcatenation_(pool_) {}
 
-  common::CompressionKind ThriftCodecToCompressionKind(
-      thrift::CompressionCodec::type codec) const {
+  static common::CompressionKind ThriftCodecToCompressionKind(
```
Comment (yingsu00) [inline velox/dwio/parquet/reader/PageReader.h:71]: This function is only used in PageReader.cpp so it doesn't need to have global scope which static functions do. Also it uses member codec_. Let's make it a private member function without parameters and move it to the private section.

Comment (nmahadevuni) [reply velox/dwio/parquet/reader/PageReader.h]: @yingsu00 Made the change. Please review.

File: velox/dwio/parquet/reader/PageReader.h [hunk 7]
```diff
@@ -68,9 +68,8 @@ class PageReader {
         chunkSize_(chunkSize),
         nullConcatenation_(pool_) {}
 
-  common::CompressionKind ThriftCodecToCompressionKind(
-      thrift::CompressionCodec::type codec) const {
-    switch (codec) {
+  common::CompressionKind thriftCodecToCompressionKind() {
```
Comment (yingsu00) [inline velox/dwio/parquet/reader/PageReader.h:71]: Why are the definitions also in the header file? Can you please move the definition to the cpp file?

Comment (yingsu00) [reply velox/dwio/parquet/reader/PageReader.h:71]: Also, does it need to be public? If not, please move the declaration to the private section. Make sure the cpp file function order is the same as the .h.

**Review by yingsu00 — APPROVED** [review]

**Comment by nmahadevuni** [comment]
@Yuhta Can you please review and merge this?

File: velox/dwio/common/compression/Compression.cpp [hunk 8]
```diff
@@ -192,14 +193,160 @@ uint64_t ZlibDecompressor::decompress(
   return destLength - zstream_.avail_out;
 }
 
-class LzoDecompressor : public Decompressor {
+class LzoAndLz4DecompressorCommon : public Decompressor {
```
Comment (Yuhta) [inline velox/dwio/common/compression/Compression.cpp:196]: Do you consider override `getDecompressedLength` to leverage compressed skipping?  It looks possible with LZ4/LZO

File: velox/dwio/dwrf/common/Compression.h [hunk 9]
```diff
@@ -84,9 +85,17 @@ static std::unique_ptr<dwio::common::BufferedOutputStream> createCompressor(
       encrypter);
 }
 
-static const CompressionOptions getDwrfOrcDecompressionOptions() {
+static const CompressionOptions getDwrfOrcDecompressionOptions(
```
Comment (Yuhta) [inline velox/dwio/dwrf/common/Compression.h:88]: No `static` function in header file, change it to `inline`. Also there is no need to `const` rvalue returned

**Pull request closed without merging**